### PR TITLE
cli abbreviations implemented

### PR DESCRIPTION
**Pull request by hugobho**
Support for abbreviated 'exo' commands implemented

Feature tests written - tests for each abbreviated command is written in their respective feature tests files; tests for 'exo add' are written under add/services/without-arguments.feature

### CODE REVIEW

File: features/add/service/without-arguments.feature [hunk 1]
```diff
@@ -1,6 +1,6 @@
 Feature: interactive scaffolding
 
-  As a developer unsure abou what the command-line arguments for the Exosphere scaffolder are
+  As a developer unsure about what the command-line arguments for the Exosphere scaffolder are
```
Comment (kevgo) [inline features/add/service/without-arguments.feature]: Thanks for fixing this!

File: features/add/service/without-arguments.feature [hunk 2]
```diff
@@ -35,5 +35,12 @@ Feature: interactive scaffolding
       > testing
       """
 
+  Scenario Outline: executing "exo add" with abbreviated commands
```
Comment (kevgo) [inline features/add/service/without-arguments.feature]: Feature files should only describe one feature. The ability to call `exo add` abbreviated is a different feature than what this file is describing (adding a new service interactively). Please move this into its own file called `abbreviations.feature`.

File: features/create_app.feature [hunk 3]
```diff
@@ -23,3 +23,15 @@ Feature: scaffolding applications
 
       services:
       """
+
+  Scenario Outline: executing "exo create" with abbreviated commands
```
Comment (kevgo) [inline features/create_app.feature]: Same here, please move this into its own file, with its own user story.

File: features/add/service/without-arguments.feature (hunk 2 above)
Comment (kevgo) [inline features/add/service/without-arguments.feature]: Cool idea btw to use a scenario outline!

File: features/steps/steps.ls [hunk 4]
```diff
@@ -188,6 +193,17 @@ module.exports = ->
   @Then /^my machine is running ExoCom$/, (done) ->
     @process.wait 'exocom  online at port', done
 
+  @Then /^the full command "([^"]*)" is executed$/ (command, done) ->
+    switch command
```
Comment (kevgo) [inline features/steps/steps.ls]: I think you can dry this up to:

```ls
expected-text = switch command
  | 'exo run'  =>  'exorun'
  | 'exo test'  =>  'exo-test'
  | 'exo setup'  =>  "exo-setup'
...
```

Behold the power of Lifescript's syntactic sugar!

File: features/run.feature [hunk 5]
```diff
@@ -30,8 +30,18 @@ Feature: running Exosphere applications
...
+    | RUN-COMMAND |
+    | exo r       |
+    | exo ru      |
+
 
   Scenario: a service crashes during startup
     Given a set-up "crashing-service" application
     When running "exo run" in this application's directory
-    Then it prints "Service 'crasher' crashed" in the terminal
+    Then it prints "Service 'crasher' crashed" in the terminal
```
Comment (kevgo) [inline features/run.feature]: Please add an empty line at the end here

File: features/setup.feature [hunk 6]
```diff
@@ -15,3 +15,15 @@ Feature: Setup of Exosphere applications
...
+    Given a set-up "tests-passing" application
+    When executing the abbreviated command <SETUP-COMMAND> in the terminal
+    Then the full command "exo setup" is executed
+
+  Examples:
+    | SETUP-COMMAND |
+    | exo s         |
+    | exo se        |
+    | exo set       |
+    | exo setu      |
```
Comment (kevgo) [inline features/setup.feature]: Let's try to remove some test cases from here that don't add much coverage. What do you think about just checking for `exo s` and `exo setu` here? If those edge cases work, its pretty plausible that the middle case works as well.

File: features/steps/steps.ls [hunk 7]
```diff
@@ -101,6 +101,11 @@ module.exports = ->
                                      cwd: app-dir,
                                      console: dim-console.console)
 
+  @When /^executing the abbreviated command ([^"]*) in the terminal$/, (command) ->
+    @process = new ObservableProcess(path.join(process.cwd!, 'bin', command),
+                                 cwd: app-dir,
+                                 console: dim-console.console)
```
Comment (kevgo) [inline features/steps/steps.ls]: Do we really need to see the output of the command in this test? If not, you can disable it with `console: off`

File: features/steps/steps.ls [hunk 8]
```diff
@@ -188,6 +193,17 @@ module.exports = ->
...
     @process.wait 'exocom  online at port', done
 
+  @Then /^the full command "([^"]*)" is executed$/ (command, done) ->
+    switch command
+      case "exo run" then expected-text = 'exorun'
+      case "exo test" then expected-text = 'exo-test'
+      case "exo setup" then expected-text = 'exo-setup'
+      case "exo create" then expected-text = 'We are about to create a new Exosphere application!'
+      case "exo add" then expected-text = 'We are about to add a new Exosphere service to the application!'
+    @process.wait expected-text, done
```
Comment (kevgo) [inline features/steps/steps.ls]: You should kill the `@process` here. Otherwise it keeps running in the background, while Cucumber already runs the next test. Then the two running `@process` might interfere with each other. Better end this step clean.

File: features/steps/steps.ls [hunk 9]
```diff
@@ -188,6 +193,17 @@ module.exports = ->
...
+
+  @Then /^I see "([^"]*)" in the terminal$/ (expected-text, done) ->
```
Comment (kevgo) [inline features/steps/steps.ls]: This step doesn't seem to be used. Can you delete it?

File: package.json [hunk 10]
```diff
@@ -6,14 +6,15 @@
     "exo": "bin/exo"
   },
   "dependencies": {
-    "async": "1.5.2",
+    "abbrev": "^1.0.9",
```
Comment (kevgo) [inline package.json]: Please remove the `^` here. We want to run against the exact version 1.0.9 here, and upgrade explicitly.

File: src/exo.ls [hunk 11]
```diff
@@ -8,7 +9,8 @@ require! {
 
 command-name = process.argv[2]
 return missing-command! unless command-name
-command-handler-path = "#{__dirname}/commands/#{command-name}/cli.js"
+command-abbrev = abbrev(command-names!)
```
Comment (kevgo) [inline src/exo.ls]: You should be able to inline the variable `command-abbrev` here.

**Comment by kevgo** [comment]
This is a great start, Hugo, and the core of this works! I have added mostly some comments to organize things a bit better. Great job!

**Comment by hugobho** [comment]
@kevgo

File: features/abbreviations.feature [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    | exo setu      |
+
+  Scenario Outline: executing "exo test" with abbreviated commands
+    Given a set-up "tests-passing" application
+    When executing the abbreviated command <TEST-COMMAND> in the terminal
+    Then the full command "exo test" is executed
+  Examples:
+    | TEST-COMMAND  |
+    | exo t         |
+    | exo tes       |
```
Comment (kevgo) [inline features/abbreviations.feature]: Add newline at the end

File: features/setup.feature [hunk 13]
```diff
@@ -15,3 +15,4 @@ Feature: Setup of Exosphere applications
       | dashboard     | node_modules |
       | mongo-service | node_modules |
       | web-server    | node_modules |
+
```
Comment (kevgo) [inline features/setup.feature]: remove this empty line

File: features/steps/steps.ls [hunk 14]
```diff
@@ -188,6 +193,16 @@ module.exports = ->
   @Then /^my machine is running ExoCom$/, (done) ->
     @process.wait 'exocom  online at port', done
 
+  @Then /^the full command "([^"]*)" is executed$/ (command, done) ->
+    switch command
+      | 'exo run' => expected-text = 'exorun'
+      | 'exo test' => expected-text = 'exo-test'
+      | 'exo setup' => expected-text = 'exo-setup'
```
Comment (kevgo) [inline features/steps/steps.ls]: Please align the `=>` with each other, so this becomes easier to read

Comment (kevgo) [inline features/steps/steps.ls]: In LiveScript everything is an expression. The switch statement returns the last value of the branch executed.

```ls
expected-text = switch command
  | 'exo run'  =>  'exorun'
  | 'exo test' =>  'exo-test'
  ...

File: features/steps/steps.ls [hunk 15]
```diff
@@ -188,6 +193,16 @@ module.exports = ->
...
+      | 'exo create' => expected-text = 'We are about to create a new Exosphere application!'
+      | 'exo add' => expected-text = 'We are about to add a new Exosphere service to the application!'
+    @process.wait expected-text, ~>
+      @process.kill!
+      done!
```
Comment (kevgo) [inline features/steps/steps.ls]: ..on 'ended', done

File: features/test.feature [hunk 16]
```diff
@@ -14,7 +14,6 @@ Feature: testing an Exosphere application
     And it prints "tweets-service works" in the terminal
     And it prints "All tests passed" in the terminal
 
-
```
Comment (kevgo) [inline features/test.feature]: Please keep two empty lines between scenarios

**Comment by kevgo** [comment]
This is ready to 🚢  once all the remaining comments have been addressed. Nicely done!
